pykafka/rdkafka/_rd_kafkamodule.c

```diff
@@ -0,0 +1,448 @@
+                                      &topic_name,
+                                      &partition_ids,
+                                      &start_offsets)) {
+        return NULL;
+    }
+
+    // We'll keep our own copy of partition_ids, because the one handed to us
+    // might be mutable, and weird things could happen if the list used on init
+    // is different than that on dealloc
+    if (self->partition_ids) {
+        set_PyRdKafkaError(RD_KAFKA_RESP_ERR__FAIL, "Already started.");
+        return NULL;
+    }
+    self->partition_ids = PySequence_List(partition_ids);
+    if (! self->partition_ids) return NULL;
+
+    // Configure and start a new RD_KAFKA_CONSUMER
+    char errstr[512];
+    self->rdk_handle = rd_kafka_new(
+            RD_KAFKA_CONSUMER, self->rdk_conf, errstr, sizeof(errstr));
+    self->rdk_conf = NULL;  // deallocated by rd_kafka_new()
+    if (! self->rdk_handle) {
+        set_PyRdKafkaError(RD_KAFKA_RESP_ERR__FAIL, errstr);
+        return NULL;
+    }
+    if (rd_kafka_brokers_add(self->rdk_handle, brokers) == 0) {
+        // XXX add brokers via conf setting instead?
+        set_PyRdKafkaError(RD_KAFKA_RESP_ERR__FAIL, "adding brokers failed");
+        goto fail;
+    }
+
+    // Configure and take out a topic handle
+    self->rdk_topic_handle =
+        rd_kafka_topic_new(self->rdk_handle, topic_name, self->rdk_topic_conf);
+    self->rdk_topic_conf = NULL;  // deallocated by rd_kafka_topic_new()
+    if (! self->rdk_topic_handle) {
+        set_PyRdKafkaError(rd_kafka_errno2err(errno), NULL);
+        goto fail;
+    }
+
+    // Start a queue and add all partition_ids to it
+    self->rdk_queue_handle = rd_kafka_queue_new(self->rdk_handle);
+    if (! self->rdk_queue_handle) {
+        set_PyRdKafkaError(RD_KAFKA_RESP_ERR__FAIL, "could not get queue");
+        goto fail;
+    }
+    Py_ssize_t i, len = PyList_Size(self->partition_ids);
+    for (i = 0; i != len; ++i) {
+        // We don't do much type-checking on partition_ids/start_offsets as this
+        // module is intended solely for use with the py class that wraps it
+        int32_t part_id = PyInt_AsLong(PyList_GetItem(self->partition_ids, i));
+        if (part_id == -1 && PyErr_Occurred()) goto fail;
+        PyObject *offset_obj = PySequence_GetItem(start_offsets, i);
+        if (! offset_obj) goto fail;  // shorter seq than partition_ids?
+        int64_t offset = PyLong_AsLongLong(offset_obj);
+        if (-1 == rd_kafka_consume_start_queue(self->rdk_topic_handle,
+                                               part_id,
+                                               offset,
+                                               self->rdk_queue_handle)) {
+            set_PyRdKafkaError(rd_kafka_errno2err(errno), NULL);
+            goto fail;
+        }
+    }
+    Py_INCREF(Py_None);
+    return Py_None;
+
+fail:  ;
+    PyObject *err_type, *err_value, *err_traceback;
+    PyErr_Fetch(&err_type, &err_value, &err_traceback);
```

Comment:
Does this line make the `goto`s above necessary? It seems like it would be cleaner to have this `fail` in its own function, but from what I can find about the `PyErr_Fetch` function, it relies on being called in the same stack frame as the error it's trying to handle. I guess not every language can be python.